File: data_worker/src/utils/circuit_breaker.py
import asyncio
from collections.abc import Awaitable, Callable
import time
from typing import Any, Literal

from data_worker.src.models.sys.config import CircuitBreakerConfig



class CircuitOpenError(RuntimeError):
    """熔断已打开时抛出的异常。"""


class CircuitBreaker:
    """简单的进程内熔断器实现（单实例内线程安全需求较低）。"""

    def __init__(self, name: str, cfg: CircuitBreakerConfig | None = None):
        self.name = name
        self.cfg = cfg or CircuitBreakerConfig()
        self._state: Literal["closed", "open", "half_open"] = "closed"
        self._failure_count: int = 0
        self._last_failure_ts: float = 0.0
        self._half_open_in_flight: int = 0
        self._lock = asyncio.Lock()

    @property
    def state(self) -> str:
        return self._state

    async def _before_call(self) -> None:
        async with self._lock:
            now = time.monotonic()
            if self._state == "open":
                # 冷却时间未到，直接拒绝
                if now - self._last_failure_ts < self.cfg.recovery_timeout:
                    raise CircuitOpenError(f"circuit '{self.name}' is open")
                # 冷却时间已到，进入 half-open
                self._state = "half_open"
                self._half_open_in_flight = 0

            if self._state == "half_open":
                if self._half_open_in_flight >= self.cfg.half_open_max_calls:
                    raise CircuitOpenError(
                        f"circuit '{self.name}' is half-open and busy"
                    )
                self._half_open_in_flight += 1

    async def _on_success(self) -> None:
        async with self._lock:
            self._failure_count = 0
            self._state = "closed"
            self._half_open_in_flight = 0

    async def _on_failure(self) -> None:
        async with self._lock:
            self._failure_count += 1
            self._last_failure_ts = time.monotonic()
            if self._failure_count >= self.cfg.failure_threshold:
                self._state = "open"
                self._half_open_in_flight = 0

    async def call(self, func: Callable[[], Awaitable[Any]]) -> Any:
        """包装异步调用：根据当前熔断状态决定是否放行或短路。"""

        await self._before_call()
        try:
            result = await func()
        except Exception:
            await self._on_failure()
            raise
        else:
            await self._on_success()
            return result
        finally:
            # HALF_OPEN 探测调用结束，释放占用
            async with self._lock:
                if self._state == "half_open" and self._half_open_in_flight > 0:
                    self._half_open_in_flight -= 1

File: certification_server/src/utils/circuit_breaker.go
package utils

import (
	modelsystem "certification_server/src/models/system"
	"context"
	"fmt"
	"sync"
	"time"
)

type CircuitBreakerState string

const (
	StateClosed   CircuitBreakerState = "closed"
	StateOpen     CircuitBreakerState = "open"
	StateHalfOpen CircuitBreakerState = "half_open"
)

// CircuitBreaker 定义了简单的熔断器结构体实现。
type CircuitBreaker struct {
	Name             string
	State            CircuitBreakerState
	cfg              modelsystem.CircuitBreakerConfig
	mu               sync.Mutex
	failureCount     int
	lastFailureTime  time.Time
	halfOpenInFlight int
}

// NewCircuitBreaker 创建一个新的 CircuitBreaker 实例。
func NewCircuitBreaker(name string, cfg *modelsystem.CircuitBreakerConfig) *CircuitBreaker {
	finalCfg := defaultCircuitBreakerConfig()
	if cfg != nil {
		if cfg.FailureThreshold > 0 {
			finalCfg.FailureThreshold = cfg.FailureThreshold
		}
		if cfg.RecoveryTimeout > 0 {
			finalCfg.RecoveryTimeout = cfg.RecoveryTimeout
		}
		if cfg.HalfOpenMaxCalls > 0 {
			finalCfg.HalfOpenMaxCalls = cfg.HalfOpenMaxCalls
		}
	}

	return &CircuitBreaker{
		Name:  name,
		State: StateClosed,
		cfg:   finalCfg,
	}
}

// Call 提供最简调用入口，适合不需要返回值的业务函数。
func (cb *CircuitBreaker) Call(ctx context.Context, fun func()) (any, error) {
	if fun == nil {
		return nil, &modelsystem.ErrCallFuncNil
	}
	return cb.CallWithResult(ctx, func(context.Context) (any, error) {
		fun()
		return nil, nil
	})
}

// CallWithResult 提供带结果和错误返回的调用入口。
func (cb *CircuitBreaker) CallWithResult(
	ctx context.Context,
	fun func(context.Context) (any, error),
) (result any, err error) {
	if cb == nil {
		return nil, &modelsystem.ErrNoCircuitBreaker
	}
	if fun == nil {
		return nil, &modelsystem.ErrCallFuncNil
	}
	if ctx == nil {
		ctx = context.Background()
	}

	if err = cb.beforeCall(time.Now()); err != nil {
		return nil, err
	}

	defer func() {
		if r := recover(); r != nil {
			cb.recordFailure(time.Now())
			err = fmt.Errorf("%w: %v", &modelsystem.ErrCircuitProtectedCallPanic, r)
			result = nil
		}
	}()

	if ctx.Err() != nil {
		cb.recordFailure(time.Now())
		return nil, ctx.Err()
	}

	result, err = fun(ctx)
	if err != nil {
		cb.recordFailure(time.Now())
		return nil, err
	}

	cb.recordSuccess()
	return result, nil
}

// CurrentState 返回熔断器当前状态。
func (cb *CircuitBreaker) CurrentState() CircuitBreakerState {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	return cb.State
}

func (cb *CircuitBreaker) beforeCall(now time.Time) error {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	switch cb.State {
	case StateOpen:
		if now.Sub(cb.lastFailureTime) < cb.cfg.RecoveryTimeout {
			return &modelsystem.ErrCircuitOpen
		}
		cb.State = StateHalfOpen
		cb.halfOpenInFlight = 0
		cb.failureCount = 0
		cb.halfOpenInFlight++
		return nil
	case StateHalfOpen:
		if cb.halfOpenInFlight >= cb.cfg.HalfOpenMaxCalls {
			return &modelsystem.ErrHalfOpenMaxCalls
		}
		cb.halfOpenInFlight++
		return nil
	default:
		return nil
	}
}

func (cb *CircuitBreaker) recordSuccess() {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	if cb.State == StateHalfOpen {
		cb.halfOpenInFlight = maxInt(0, cb.halfOpenInFlight-1)
		// 简易实现: 半开探测一旦成功，直接恢复闭合。
		cb.State = StateClosed
		cb.failureCount = 0
		cb.lastFailureTime = time.Time{}
		cb.halfOpenInFlight = 0
		return
	}

	cb.failureCount = 0
}

func (cb *CircuitBreaker) recordFailure(now time.Time) {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	if cb.State == StateHalfOpen {
		cb.halfOpenInFlight = maxInt(0, cb.halfOpenInFlight-1)
		cb.State = StateOpen
		cb.failureCount = cb.cfg.FailureThreshold
		cb.lastFailureTime = now
		return
	}

	cb.failureCount++
	if cb.failureCount >= cb.cfg.FailureThreshold {
		cb.State = StateOpen
		cb.lastFailureTime = now
	}
}

func defaultCircuitBreakerConfig() modelsystem.CircuitBreakerConfig {
	return modelsystem.CircuitBreakerConfig{
		FailureThreshold: 5,
		RecoveryTimeout:  10 * time.Second,
		HalfOpenMaxCalls: 1,
	}
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}
